### ISSUE TITLE
`SideSetsBetweenSubdomainsGenerator` calls `mesh->find_neighbors();` every time

### ISSUE THREAD

**Issue by dschwen** [idaholab/moose#25378]
## Reason
Simulations with many `SideSetsBetweenSubdomainsGenerator`s spend a large chunk of their setup time in `UnstructuredMesh::find_neighbors()`. After the first call we should only rerun find neighbors if the mesh actually changed.

## Design
Currently we only have `mesh->is_prepared()` as a state check, but the mesh won;t be fully prepared after just running `find_neighbors`. Maybe we can add a state that is set to true when neighbors are initialized and invalidated when the mesh topology is changed. (this might have to be a libmesh thing)

## Impact
Improve setup performance.

**Comment by lindsayad** [comment]
Yes this should be a libMesh ticket. We have talked a lot about having more fine grained state checks but haven't really implemented much as it can be fairly complex to gauge state given that we give out non-const access to individual elements and nodes, etc.

Pinging @roystgnr

**Comment by dschwen** [comment]
ad59b35 is a workaround. it allows MeshGenerators to check if the preceding generators were of a certain type. Here, if the generator that ran before the current `SideSetsBetweenSubdomainsGenerator` was also a `SideSetsBetweenSubdomainsGenerator` we can safely assume that find neighbors was run and doesn't need to be run again.
